// stitch/config.js

(function (exports) {

var util = require('stitch/util'),
    ProtoClass = require('stitch/proto-class'),
    Module = require('stitch/module').Class,
    Config
;

Config = ProtoClass.create({
    
    configurations: {},
    
    get: function (name) {
        return this.configurations[name];
    },
    
    has: function (name) {
        return this.configurations.hasOwnProperty(name);
    }
    
}, {
    init: function (name) {
        this.name = name;
        this.sourcePaths = [];
        this.modules = {};
        this.filters = {};
        Config.configurations[name] = this;
    },
    
    enhance: function (fn) {
        fn.call(this, this);
        return this;
    },
    
    set: function (key, value) {
    
    },

    get: function (key) {
    
    },
    
    description: function (txt) {
        if (this.currentDesc === undefined) {
            this.currentDesc = txt;
        }
        else {
            this.currentDesc += "\n" + txt;
        }
        return this;
    },
    
    desc: function (txt) {
        return this.description(txt);
    },

    compose: function (name) {
        var mod = this.modules[name];
        return mod.compose();
    },
    
    render: function (name, type) {
        return this.compose(name).render(type);
    },
    
    module: function (name, fn) {
        var mod,
            desc
        ;
        
        if (!fn) {
            return this.getModule(name);
        }
        
        desc = this.currentDesc;
        mod = this.currentModule = this.modules[name] ||
            (this.modules[name] = new Module(name, desc));
        
        fn.call(mod, mod);

        delete this.currentDesc;
        delete this.currentModule;
        
        return this;
    },

    filter: function (type, name, fn) {
        
        return this;
    },

    getModule: function (name) {
        var mod = this.modules[name];
        
        if (!mod) {
            throw("Unknown Module '" + name + "'");
        }
        
        return mod;
    },

    getFilter: function (name) {
        name = this.mimeTypes.hasOwnProperty(name) ?
            this.mimeTypes[name] :
            name
        ;
    
        return this.filters[name];
    }
    
});

// export it
exports.Class = Config;
    
}(exports));
